app/uk/gov/hmrc/minorentityidentificationfrontend/forms/trustForms/CaptureCHRNForm.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.minorentityidentificationfrontend.forms.trustForms
+
+import play.api.data.Form
+import play.api.data.Forms.text
+import play.api.data.validation.Constraint
+import uk.gov.hmrc.minorentityidentificationfrontend.forms.utils.ConstraintUtil.ConstraintUtil
+import uk.gov.hmrc.minorentityidentificationfrontend.forms.utils.ValidationHelper.{validate, validateNot}
+
+import scala.util.matching.Regex
+
+object CaptureCHRNForm {
+
+  val ChrnKey: String = "chrn"
+
+  val ChrnNotEnteredErrorKey: String = "chrn.error_not_entered"
+  val ChrnInvalidCharactersErrorKey: String = "chrn.error_invalid_characters"
+  val ChrnInvalidLengthErrorKey: String = "chrn.error_invalid_length"
+
+  val ChrnRegex: Regex = "[A-Z]{1,2}[0-9]{1,5}".r
+
+  def chrnEmpty: Constraint[String] = Constraint("chrn.not_entered")(
+    chrn => validate(constraint = chrn.isEmpty, errMsg = ChrnNotEnteredErrorKey)
+  )
+
+  def chrnInvalidCharacters: Constraint[String] = Constraint("chrn.invalid_characters")(
+    chrn => validateNot(
+      constraint = chrn matches ChrnRegex.regex,
+      errMsg = ChrnInvalidCharactersErrorKey
+    )
+  )
+
+  def chrnInvalidLength: Constraint[String] = Constraint("chrn.invalid_length")(
+    chrn => validate(
+      constraint = chrn.length != 7,
+      errMsg = ChrnInvalidLengthErrorKey
+    )
+  )
+
+  val form: Form[String] =
+    Form(
+      ChrnKey -> text.verifying(chrnEmpty andThen chrnInvalidLength andThen chrnInvalidCharacters)
```

Comment:
If you put in a string less than 7 in length but with invalid characters it shows 'Enter a HMRC reference number that is 7 characters or less' maybe swap the last two round here so it does the invalid characters one first?
chrnEmpty andThen chrnInvalidCharacters andThen chrnInvalidLength

Reply:
There is a complication here in that the regex used for invalid characters also implicitly checks the length of the CHRN. As such if we put the invalid character check before the length check we will never see the invalid length error. So we either keep the checks as they are or perhaps look at using a mapping function which may give us more options as we did with the Utr yesterday.